File: src/pipeline/dag.ts
import { PipelineStage, StageDependency } from './types.js';

/**
 * Directed acyclic graph for pipeline stages.
 * Provides topological sorting and identifies parallelizable stage groups.
 */
export class PipelineDAG {
  private deps: Map<PipelineStage, Set<PipelineStage>> = new Map();
  private stages: Set<PipelineStage> = new Set();

  constructor(dependencies: StageDependency[]) {
    for (const { stage, dependsOn } of dependencies) {
      this.stages.add(stage);
      this.deps.set(stage, new Set(dependsOn));
      for (const dep of dependsOn) {
        this.stages.add(dep);
        if (!this.deps.has(dep)) {
          this.deps.set(dep, new Set());
        }
      }
    }
    this.validateNoCycles();
  }

  /**
   * Returns stages grouped by execution level.
   * Stages within the same group have no dependencies on each other
   * and can run in parallel.
   *
   * Example for PLAN->BUILD->TEST->REVIEW:
   *   [[PLANNING], [BUILDING], [TESTING], [REVIEWING]]
   *
   * Example with LINTING depending only on PLANNING:
   *   [[PLANNING], [BUILDING, LINTING], [TESTING], [REVIEWING]]
   */
  getExecutionLevels(): PipelineStage[][] {
    const levels: PipelineStage[][] = [];
    const completed = new Set<PipelineStage>();
    const remaining = new Set(this.stages);

    while (remaining.size > 0) {
      const level: PipelineStage[] = [];

      for (const stage of remaining) {
        const deps = this.deps.get(stage) || new Set();
        const allDepsCompleted = [...deps].every(d => completed.has(d));
        if (allDepsCompleted) {
          level.push(stage);
        }
      }

      if (level.length === 0) {
        throw new Error('DAG has unresolvable dependencies — possible cycle or missing stage');
      }

      // Sort within level for deterministic ordering
      level.sort();

      for (const stage of level) {
        remaining.delete(stage);
        completed.add(stage);
      }

      levels.push(level);
    }

    return levels;
  }

  /**
   * Get the flat topological order (for backward compatibility).
   */
  getTopologicalOrder(): PipelineStage[] {
    return this.getExecutionLevels().flat();
  }

  /**
   * Get direct dependencies for a stage.
   */
  getDependencies(stage: PipelineStage): PipelineStage[] {
    return [...(this.deps.get(stage) || [])];
  }

  /**
   * Check if running the given subset of stages is valid
   * (all dependencies for each stage are either in the subset or skipped).
   */
  filterStages(subset: PipelineStage[]): StageDependency[] {
    const subsetSet = new Set(subset);
    return subset.map(stage => ({
      stage,
      dependsOn: [...(this.deps.get(stage) || [])].filter(d => subsetSet.has(d)),
    }));
  }

  private validateNoCycles(): void {
    const visited = new Set<PipelineStage>();
    const inStack = new Set<PipelineStage>();

    const dfs = (stage: PipelineStage) => {
      if (inStack.has(stage)) {
        throw new Error(`Cycle detected in pipeline DAG involving stage: ${stage}`);
      }
      if (visited.has(stage)) return;
      visited.add(stage);
      inStack.add(stage);
      for (const dep of this.deps.get(stage) || []) {
        dfs(dep);
      }
      inStack.delete(stage);
    };

    for (const stage of this.stages) {
      dfs(stage);
    }
  }
}

File: backend/src/haan/pipeline/dag.py
"""
DAG (Directed Acyclic Graph) computation for pipeline stage ordering.

Provides topological sort and parallel execution group computation
so the pipeline engine knows which stages can run concurrently and
which must wait for dependencies.
"""

from __future__ import annotations

from collections import defaultdict, deque

from haan.pipeline.types import PipelineStage, StageDependency


class PipelineDAG:
    """
    Directed Acyclic Graph for pipeline stage dependencies.

    Computes execution order using Kahn's algorithm (topological sort)
    and groups stages into parallel execution levels.
    """

    def __init__(self, dependencies: list[StageDependency]) -> None:
        """
        Initialize the DAG from a list of stage dependencies.

        Args:
            dependencies: List of StageDependency objects defining the graph edges.
        """
        self._deps = dependencies
        self._adjacency: dict[PipelineStage, list[PipelineStage]] = defaultdict(list)
        self._in_degree: dict[PipelineStage, int] = {}

        # Build the graph
        for dep in dependencies:
            if dep.stage not in self._in_degree:
                self._in_degree[dep.stage] = 0
            for parent in dep.depends_on:
                self._adjacency[parent].append(dep.stage)
                self._in_degree[dep.stage] = self._in_degree.get(dep.stage, 0) + 1
                if parent not in self._in_degree:
                    self._in_degree[parent] = 0

    def topological_sort(self) -> list[PipelineStage]:
        """
        Compute a topological ordering of the stages.

        Returns:
            List of PipelineStage values in valid execution order.

        Raises:
            ValueError: If the graph contains a cycle.
        """
        in_degree = dict(self._in_degree)
        queue: deque[PipelineStage] = deque()
        result: list[PipelineStage] = []

        # Start with stages that have no dependencies
        for stage, degree in in_degree.items():
            if degree == 0:
                queue.append(stage)

        while queue:
            stage = queue.popleft()
            result.append(stage)

            for child in self._adjacency.get(stage, []):
                in_degree[child] -= 1
                if in_degree[child] == 0:
                    queue.append(child)

        if len(result) != len(self._in_degree):
            raise ValueError(
                "Pipeline DAG contains a cycle. "
                f"Processed {len(result)} stages out of {len(self._in_degree)}."
            )

        return result

    def get_execution_levels(self) -> list[list[PipelineStage]]:
        """
        Group stages into parallel execution levels.

        Stages within the same level have no dependencies on each other
        and can be executed concurrently.

        Returns:
            List of lists, where each inner list is a group of stages
            that can run in parallel. Groups are ordered by dependency level.
        """
        in_degree = dict(self._in_degree)
        levels: list[list[PipelineStage]] = []

        # Start with all stages that have no dependencies
        current_level = [
            stage for stage, degree in in_degree.items() if degree == 0
        ]

        while current_level:
            levels.append(sorted(current_level, key=lambda s: s.value))
            next_level: list[PipelineStage] = []

            for stage in current_level:
                for child in self._adjacency.get(stage, []):
                    in_degree[child] -= 1
                    if in_degree[child] == 0:
                        next_level.append(child)

            current_level = next_level

        return levels

    def filter_stages(self, stages: list[PipelineStage]) -> list[StageDependency]:
        """
        Filter the DAG to only include the specified stages.

        Dependencies that reference excluded stages are removed.

        Args:
            stages: List of stages to keep.

        Returns:
            Filtered list of StageDependency objects.
        """
        stage_set = set(stages)
        return [
            StageDependency(
                stage=dep.stage,
                depends_on=[d for d in dep.depends_on if d in stage_set],
            )
            for dep in self._deps
            if dep.stage in stage_set
        ]

    def get_dependencies(self, stage: PipelineStage) -> list[PipelineStage]:
        """Return the direct dependencies of a stage."""
        for dep in self._deps:
            if dep.stage == stage:
                return list(dep.depends_on)
        return []
